esp/esp/program/modules/handlers/usergroupmodule.py

```diff
@@ -0,0 +1,124 @@
+    doc = """Modify which users are in which user groups"""
+
+    @classmethod
+    def module_properties(cls):
+        return {
+            "admin_title": "User Group Module",
+            "link_title": "Manage User Groups",
+            "module_type": "manage",
+            "seq": 501,
+            "choosable": 1,
+        }
+
+    @aux_call
+    @needs_admin
+    def usergroupfinal(self, request, tl, one, two, module, extra, prog):
+        if request.method != 'POST' or 'filterid' not in request.GET or (request.POST.get('group_name_new', '') and request.POST.get('group_name_old', '')):
+            raise ESPError(), 'Filter and/or group has not been properly set'
+
+        # get the filter to use and text message to send from the request; this is set in grouptextpanel form
+        filterObj = PersistentQueryFilter.objects.get(id=request.GET['filterid'])
+        if request.POST.get('group_name_new', ''):
+            group = request.POST['group_name_new']
+        else:
+            group = request.POST['group_name_old']
+        clean = False
+        if 'group_clean' in request.POST:
+            clean = request.POST['group_clean']
+
+        log = self.updateGroup(filterObj, group, clean)
+
+        return render_to_response(self.baseDir()+'finished.html', request, {'log': log})
+
+    @main_call
+    @needs_admin
+    def usergroup(self, request, tl, one, two, module, extra, prog):
+        usc = UserSearchController()
+        context = {}
+        context['program'] = prog
+
+        if request.method == "POST":
+            data = ListGenModule.processPost(request)
+            filterObj = UserSearchController().filter_from_postdata(prog, data)
+
+            context['filterid'] = filterObj.id
+            context['num_users'] = ESPUser.objects.filter(filterObj.get_Q()).distinct().count()
+            context['groups'] = Group.objects.all().values_list('name', flat=True)
+            return render_to_response(self.baseDir()+'options.html', request, context)
+
+        context.update(usc.prepare_context(prog, target_path='/manage/%s/usergroup' % prog.url))
+        return render_to_response(self.baseDir()+'search.html', request, context)
+
+    @staticmethod
+    def updateGroup(filterobj, group_name, clean = False):
+        """ Adds the specified users to the specified group. Removes anyone else if clean == True.
+            Returns a log which can be displayed to user. """
+
+        users = filterobj.getList(ESPUser)
+        try:
+            users = users.distinct()
+        except:
+            pass
+
+        if not users:
+            raise ESPError(), "Your query did not match any users"
+
+        group, created = Group.objects.get_or_create(name = group_name)
+
+        if clean:
+            group.user_set.clear()
+
+        group.user_set.add(*users)
+
+        message = "%i users have been added to user group %s. User group %s now has %i users." % (users.count(), group.name, group.name, group.user_set.count())
```

Comment:
It might be helpful to do something similar to what we do with checkins, and state how many of the users we tried to add were ALREADY in the group, how many we added, and how many there are total?